### ISSUE TITLE
arm_neon.h intrinsics should be target-gated, not preprocessor-gated

### ISSUE THREAD

**Issue by davidben** [llvm/llvm-project#56480]
Clang's intrinsics headers on Arm contain code like:

```cpp
#if !defined(__ARM_NEON)
#error "NEON support not enabled"
#else
```
or:
```cpp
#if __ARM_ARCH >= 8 && defined(__ARM_FEATURE_AES)
#ifdef __LITTLE_ENDIAN__
__ai uint8x16_t vaesdq_u8(uint8x16_t __p0, uint8x16_t __p1) {
...
```

(Generated by https://github.com/llvm/llvm-project/blob/main/clang/utils/TableGen/NeonEmitter.cpp.)

This means that one can only use Arm intrinsics in TUs that mark the feature as available for the entire intrinsic, e.g. via `-march` flags. In contrast, the x86 intrinsics are consistently defined, but tagged with `__attribute__((__target__("whatever")))`:
https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/avx2intrin.h#L18

The x86 variant is both easier to use, as it doesn't require messing with your project's build definition, and safer, since you can just write a `target("avx2")` function and then gate the unmarked -> marked transition on some suitable CPUID check. (Or perhaps even use [multi-versioning](https://gcc.gnu.org/onlinedocs/gcc/Function-Multiversioning.html), though I believe target attributes are usable even without that.)

In contrast, Clang's Arm story requires messing with build definitions and risks ODR violations. Suppose your NEON-enabled TU included some inline functions, where Clang happened to vectorize code and use NEON instructions. If that copy of the inline function won, the resulting binary would inadvertently require NEON, even if the overall target wasn't meant to require NEON.

GCC fixed their Arm intrinsics, back in 2015, to be target-gated instead. Arm intrinsics would be much more usable if Clang could get parity here.
https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;h=ae5e29239e28818f807cf11775c95c4243d9a256;hp=b8c7c62b2dbbdf355adb56d8250e68222ae0febb

**Comment by davidben** [comment]
Playing around with godbolt, it looks like making this work for NEON itself may be more involved than just fixing the header:
```
<source>:5:24: error: 'neon_vector_type' attribute is not supported on targets missing 'neon' or 'mve'; specify an appropriate -march= or -mcpu=
typedef __attribute__((neon_vector_type(4))) uint32_t uint32x4_t;
```
Not sure about other features, which don't depend on feature-gated types.

**Comment by jan-wassenberg** [comment]
+1, this is important for SVE and even NEON adoption.
Here's a [Compiler Explorer example](https://gcc.godbolt.org/z/Yobn3vdf6) of this in action using GCC.

**Comment by nico** [comment]
Seems like a cool feature.

I also ran into this while trying to write a faster SHA256 using intrinsics over at #56121.

Adding a random assortment of folks who touched arm_neon.td and arm_sve.td: @davemgreen @dcandler @sdesmalen-arm , wdyt?

@davidben linked to the GCC change already.

9fc7fb274e5341b14c71487cf413343439581d50 is the change that did the corresponding change for the intel intrinsics long ago.

**Comment by ktkachov-arm** [comment]
This looks somewhat related to the function multiversioning work from @ilinpv  https://reviews.llvm.org/D127812 but perhaps is not blocked by it?

**Comment by davidben** [comment]
Yeah, I think they're related but can be done independently. Doing this would make function multiversioning much more useful (otherwise function multiversioning is limited to compiler vectorization, as I understand), but function multiversioning is not necessary for this to be useful, since the application can always manage the dispatch itself.

**Comment by davemgreen** [comment]
OK that's good. Thanks for clarifying. It is something that seems worth fixing in general, as the difference between compilers is likely to trip some people up.

**Comment by jan-wassenberg** [comment]
Sure, all other things being equal, more compatibility across compilers is nice :)

**Comment by jan-wassenberg** [comment]
That's fantastic news, thanks for the heads-up!
Yes, I believe changing HWY_HAVE_RUNTIME_DISPATCH to 1 would be sufficient. If the tests outputs include [..]/NEON and [..]/SVE in the same binary, without setting -march to +sve, then we've succeeded.

I'll be happy to double-check this when back in early Feb, presumably this will require building clang from source.

Should we then update the Highway code to expect this capability in Clang 16?

**Comment by Lukacma** [comment]
This issue has been fixed in #95224.
